### PR TITLE
Replace each_allocated_chunk function with iter_allocated_chunks function

### PR DESCRIPTION
**Pull request by TethysSvensson**
This implements the suggestion in #27.

### CODE REVIEW

**Comment by TethysSvensson** [comment]
Note that this currently removes the function `each_allocated_chunk`, since I believe this new functionality to be strictly superior. The iterator API is more usable, while also being safe to use as long as you do not look at any of the byte values. The type signature makes it clear exactly which parts are unsafe.

I can easily re-implement the old function on top of the iterator for backwards-compatiblity, if you prefer. If you want me to do this, I think we should consider marking as deprecated, or perhaps changed to use `[MaybeUninit<u8>]` slices instead of `[u8]` slices (though that you also be a breaking change).

**Comment by TethysSvensson** [comment]
By using google and github search, I have not been able to find any references to `each_allocated_chunk` outside of this crate or rustdocs. There might be users of the function in private or non-indexed code though.

**Comment by fitzgen** [comment]
> By using google and github search, I have not been able to find any references to `each_allocated_chunk` outside of this crate or rustdocs. There might be users of the function in private or non-indexed code though.

I use it here, fwiw: https://github.com/fitzgen/dodrio/blob/22faf1eb60f9bcb16094e1b2f4575edfb203338e/src/change_list/emitter.rs#L43

File: src/lib.rs [hunk 1]
```diff
@@ -663,10 +667,10 @@ impl Bump {
         }
     }
 
-    /// Call `f` on each chunk of allocated memory that this arena has bump
-    /// allocated into.
+    /// Returns an iterator over the each chunk of allocated memory that
```
Comment (fitzgen) [inline src/lib.rs]:
```suggestion
    /// Returns an iterator over each chunk of allocated memory that
```

File: src/lib.rs [hunk 2]
```diff
@@ -663,10 +667,10 @@ impl Bump {
...
+    /// from an arena has bump allocated into.
```
Comment (fitzgen) [inline src/lib.rs]:
```suggestion
    /// this arena has bump allocated into.
```

File: src/lib.rs [hunk 3]
```diff
@@ -705,20 +710,39 @@ impl Bump {
...
-            let foot = foot.as_ref();
+    pub fn iter_allocated_chunks(&mut self) -> ChunkIter<'_> {
+        ChunkIter {
+            footer: Some(self.all_chunk_footers.get()),
+            bump: PhantomData,
+        }
+    }
+}
 
+/// An iterator over the each chunk of allocated memory that
```
Comment (fitzgen) [inline src/lib.rs]:
```suggestion
/// An iterator over each chunk of allocated memory that
```

File: src/lib.rs [hunk 4]
```diff
@@ -705,20 +710,39 @@ impl Bump {
...
+/// from an arena has bump allocated into.
```
Comment (fitzgen) [inline src/lib.rs]:
```suggestion
/// an arena has bump allocated into.
```

File: src/lib.rs [hunk 5]
```diff
@@ -705,20 +710,39 @@ impl Bump {
...
+///
+/// Chunks are returned in order of allocation: oldest chunks first, newest
+/// chunks last.
+///
+/// This struct is created by the [`iter_allocated_chunks`] method on
+/// [Bump]. See that function for a safety description regarding.
```
Comment (fitzgen) [inline src/lib.rs]:
```suggestion
/// [Bump]. See that function for a safety description regarding reading from the returned items.
```

**Comment by TethysSvensson** [comment]
> > By using google and github search, I have not been able to find any references to `each_allocated_chunk` outside of this crate or rustdocs. There might be users of the function in private or non-indexed code though.
> 
> I use it here, fwiw: https://github.com/fitzgen/dodrio/blob/22faf1eb60f9bcb16094e1b2f4575edfb203338e/src/change_list/emitter.rs#L43

That also shows up on github search now, but didn't before. It still does not show up on google search. This makes me a lot less inclined to trust github search in the future.


> Can you also reimplement `each_allocated_chunk` in terms of `iter_allocated_chunks` with a `#[deprecated(note = "deprecated in favor of iter_allocated_chunks")]` note?

Done.

**Pull request merged**